Addressing docker comments

bam241 opened the pull request:
This address comments from #660 about Docker.


I added something from me as well, see if I can separate the housekeeping jobs from the others... and that way simply the travis CI file

bam241 committed:
adress most of the comments

bam241 committed:
move sphinx to 18.04 only -- not working for 16.04..., and test a jobs for housekeeping

bam241 committed:
removing housekeeping from matrix

bam241 committed:
is this working ?

bam241 committed:
is this working ?

bam241 committed:
is this working ?

bam241 committed:
back

bam241 committed:
add stage?

bam241 committed:
add stage?

bam241 committed:
good syntax?

bam241 committed:
good syntax? bis

bam241 committed:
good syntax? ter

bam241 committed:
good syntax? quad

bam241 committed:
new test

bam241 committed:
this should work better

bam241 committed:
this should work better

bam241 committed:
better with diff indent?

bam241 committed:
better with other formating?

bam241 committed:
better with other formating?

bam241 committed:
better with other formating?

bam241 committed:
better with other formating?

bam241 committed:
cleaning syntax

bam241 committed:
testintg new syntax

bam241 committed:
testintg new syntax

bam241 committed:
removing housekeeping tool from docker

bam241 committed:
add new

bam241 committed:
testing install pshinx

bam241 committed:
uncomment push

bam241 committed:
cleaning

bam241 commented:
@kkiesling @pshriwise @gonuke this is ready for review.

pshriwise commented on docker/build_geant4.sh:
```diff
@@ -1,13 +1,13 @@
 mkdir -p ${geant4_build_dir}/bld
 cd ${geant4_build_dir}
-wget http://geant4.cern.ch/support/source/geant4.${geant4_version}.tar.gz
+wget https://geant4.cern.ch/support/source/geant4.${geant4_version}.tar.gz --no-check-certificate
```
In #660 they mention verifying our file against a checksum. Where do we stand on adding that?

bam241 replied:
did not found any checksum for geant4 tarball :(

gonuke replied:
We could generate a checksum against the version we are pulling

bam241 replied:
make sense !

pshriwise reviewed (commented):
Just one small comment from me. Nice cleanup of the housekeeping check!

ljacobson64 commented on docker/Dockerfile:
```diff
@@ -2,30 +2,18 @@ ARG UBUNTU_VERSION=18.04
-
 # Get some pip packages
-RUN pip install cython sphinx
+RUN pip install cython
```
Why was sphinx moved out of this file?

kkiesling replied:
It looks like anything having to do w/ housekeeping was moved to the housekeeping.sh file.

ljacobson64 commented on docker/Dockerfile:
```diff
@@ -2,30 +2,18 @@ ARG UBUNTU_VERSION=18.04
                        libhdf5-dev liblapack-dev libpython2.7-dev libtool \
                        python-numpy python-pip python-setuptools wget
 
-# Get astyle
```
Why was astyle moved out of this file? Everything that could possibly mess with system libraries (i.e. apt or pip) was put in this file.

bam241 replied:
I put it back in :)

ljacobson64 commented on docker/build_dagmc.sh:
```diff
@@ -13,7 +13,6 @@ fi
 rm -rf ${dagmc_build_dir}/bld ${dagmc_install_dir}
 mkdir -p ${dagmc_build_dir}/bld
 cd ${dagmc_build_dir}
-#git clone https://github.com/svalinn/DAGMC -b develop
```
The reason these lines were commented is that they would normally be part of a DAGMC build script, but in this specific case, we get the DAGMC source code by copying it into the docker image, not cloning it from github.

kkiesling replied:
@ljacobson64 is there any reason to leave this line in as a comment w/ justification or can we just remove it all together? (see [here](https://github.com/svalinn/DAGMC/pull/660#discussion_r368664684))

bam241 replied:
If I understand correctly, there are there just because the normal install `DAGMC` procedure should include them. But in this case as we copy the `DAGMC` from the PR (using `Travis`) we don't need them.

If this is correct, I don't think they should be there:
Those scripts are dedicated for CI not to show people how to install `DAGMC`.
It is probably a good idea to have a version of them somewhere to facilitate `DAGMC` install but for me this is definitely not the place.

kkiesling replied:
I agree w/ this assessment @bam241. I think the lines here should be removed and if we want a docker to help users, we need to host that elsewhere.

ljacobson64 replied:
For maintainability purposes, I think it would be good to have some indication of why the source code already exists in this directory. (The current situation with just a commented `git clone` line is also not very helpful.) Perhaps a comment line that says

```
# DAGMC source code already exists in this directory because it was copied into
# the docker image by travis.yml
```

bam241 replied:
I think I understand,

the line is left there to highlight that we don't need to copy DAGMC source, while we normally would in a standard installation process.
So the line is basically there for educational purposes...
Is that correct @ljacobson64 ?

I would definitely prefer to not mix purposes in those scripts. So if there are used for CI install and testing, I would remove the git clone line and add the comment at the head of the file.

ljacobson64 commented on docker/build_dagmc.sh:
```diff
@@ -29,4 +28,3 @@ cmake ../src -DMOAB_DIR=${moab_install_dir} \
 make -j${jobs}
 make install
 cd
-#rm -rf ${dagmc_build_dir}
```
The dagmc_build_dir would normally be deleted after building DAGMC, but we need to keep it in this specific case in order to run the tests.

kkiesling replied:
My same question above applies here too- do you think it should be left in as a comment w/ an explanation or removed?

bam241 replied:
@ljacobson64 

agreed with @kkiesling if this is commented without explanation it should not be part of the build.
I'd be happy to add them back with a line of explanation. 

when you mean run the tests, do you mean run the test locally ? 
Because we run the test in CI without those...

ljacobson64 replied:
I think this line can be deleted.

To answer @bam241's question: I do mean for running the tests in CI. The CI runs `make test` in the DAGMC build directory, so the DAGMC build directory must exist in order to run the tests. See here: https://github.com/bam241/DAGMC/blob/fix_docker/docker/travis.tests.sh#L28

ljacobson64 commented on .travis.yml:
```diff
@@ -66,17 +34,49 @@ before_install:
+    - env: UBUNTU_VERSION=18.04 COMPILER=clang HDF5_VERSION=1.10.4 MOAB_VERSION=develop
+      if: type != pull_request
+    - stage: housekeeping
+      env: UBUNTU_VERSION=16.04 COMPILER=gcc   HDF5_VERSION=1.10.4 MOAB_VERSION=5.1.0
```
The housekeeping run should use 18.04

bam241 replied:
oups this was not intended

ljacobson64 commented on docker/Dockerfile:
```diff
@@ -2,30 +2,18 @@ ARG UBUNTU_VERSION=18.04
                        python-numpy python-pip python-setuptools wget
 
-# Get astyle
-RUN if [ "${UBUNTU_VERSION}" == "18.04" ]; then \
```
Removing these lines will disallow the ability to do the housekeeping run on Ubuntu 16.04. This is fine, as long as the housekeeping run is changed back to being on 18.04.

bam241 replied:
I re-added those line for ubuntu 18.04 only as housekeeping is only done there.

kkiesling reviewed (approved):
I have no major comments. Thanks @bam241!

bam241 committed:
change ubuntu version for housekeeping

bam241 commented:
@ljacobson64 to avoid having the housekeeping stuff everywhere and only used for 1 job in ubuntu 18.04. I moved `astyle` and `sphinx` install in the housekeeping scripts.

bam241 committed:
test simplifying docker commits

bam241 committed:
change commit to container...

bam241 committed:
not all of them are suppose to be container...

bam241 committed:
add comments

bam241 committed:
run_in_docker alias

bam241 committed:
alias did not work

bam241 committed:
new to get the container id again?

gonuke commented on docker/travis.housekeeping.sh:
```diff
@@ -18,6 +18,9 @@ if [ "${TRAVIS_REPO_SLUG}" == "svalinn/DAGMC" ] && \
 fi
 
+# install package for style check
+apt-get install astyle
```
Why are we moving this to here?

gonuke commented on docker/travis.tests.sh:
```diff
@@ -19,7 +19,7 @@ cd ${dagmc_build_dir}/bld
 if [ "${TRAVIS_PULL_REQUEST}" == "false" ]; then
   cd src/make_watertight/tests
-  wget ${MW_REG_TEST_MODELS_URL} -O mw_reg_test_files.tar.gz -o wget.out
+  wget ${MW_REG_TEST_MODELS_URL} -O mw_reg_test_files.tar.gz
```
The comment suggested streaming these to `stderr` and `stdout` - do we do this now?  Or is that default behavior?

bam241 replied:
my understanding of the comment was the default behavior streams them to `stderr`

gonuke commented:
I'm ambivalent on the housekeeping changes.  They may happen fairly often, which means a lot of `apt install` and `pip installing` for stuff we could do once.  Should we have different docker images? Would that help?

bam241 committed:
moving housekeeping install in 19.04 only

bam241 committed:
adding geant4 shasum

bam241 commented:
I was thinking @ljacobson64 @kkiesling @pshriwise @gonuke 
For separation of concern what do you think about:
- creating a `.travis_scripts` folder containing all the scripts required for CI (exept the `update_docker.sh` one)
- keep only the `Dockerfile` and the `update_docker.sh` macro in the `docker` folder
- create a `script` folder, with a version of the install scripts dedicated to installation for users ?

ljacobson64 commented:
That kind of sounds like overkill. I'd rather just delete those 2 lines we were discussing

bam241 committed:
add comment about how DAGMC src got there, rm symb link as they do not seem to be necessary

bam241 commented:
> That kind of sounds like overkill. I'd rather just delete those 2 lines we were discussing

Yes but I feel like it will clarify the purpose of everything.
It is just that it does not seems obvious to me, that the scripts used to install and run the tests in CI would be located in the `docker` folder

ljacobson64 commented:
Would it make more sense if the `docker` folder were renamed `CI`?

bam241 committed:
shuffling folder around

bam241 committed:
update news

bam241 commented:
@gonuke @kkiesling @ljacobson64 this should  be ready for an around of review

bam241 committed:
fixing parsing version number

bam241 committed:
some more shuffling

bam241 committed:
forgot to uncomment

bam241 committed:
finalize change from .travis_scripts to scripts

bam241 commented:
@ljacobson64 @kkiesling I think I am now satisfied with the way it looks. 

cc @pshriwise @gonuke

bam241 committed:
reset DAGMC.cpp, this is not what this branch is about

gonuke commented:
Should we test some of this by adding an automatic build to dockerhub?

ljacobson64 commented:
How does that work?

gonuke commented:
You point dockerhub at a Dockerfile in a github repo and everytime that repo changes, it rebuilds the image.  We have deployed it for 
* https://hub.docker.com/repository/docker/svalinn/pymoab-py2-18.04
* https://hub.docker.com/repository/docker/svalinn/pymoab-py2-16.04
* https://hub.docker.com/repository/docker/svalinn/pymoab-visit-py2-18.04
and it seems to work well.

So far they have all been for images whose configuration is 100% defined in the dockerfile, so I don't know how it will behave for this image that requires scripts...

bam241 commented:
@gonuke this seems like a good idea, will take a look at this

gonuke commented:
I have something running here: https://hub.docker.com/repository/docker/svalinn/dagmc-ci-ubuntu-18.04/builds

gonuke commented:
The dockerhub build is taking a long time....  probably not surprising

gonuke commented:
Looks like it might have timed out...

bam241 commented:
@gonuek if we want to do the automated one, we will probably need 2 dockerfiles one for each ubuntu version

ljacobson64 commented:
We could stop supporting 16.04 since it's 4 years old now

ljacobson64 commented:
Although in that case we may as well start supporting 20.04...

bam241 committed:
add automated build file for dockerhub

bam241 committed:
cleaning

bam241 committed:
moving hooks in docker folder

bam241 commented:
@gonuke I think we will have problem....
timeout limits is set to 2h
https://success.docker.com/article/what-are-the-current-resource-limits-placed-on-automated-builds

bam241 commented:
I think the main problem is building geant4... that takes forever....

gonuke commented:
And we build Geant4 four different times!  It times out on the first try.

gonuke commented:
This may be a good reason to separate out some of the docker `RUN` lines, and build Geant4 before HDF5.  It is less likely to change in ways that matter, and we can do that locally, push the image and then let the autobuild skip the parts that are already built.

That said, perhaps we make a bunch of different images, each tagged with the info we we want and let dockerhub build them all?

TL;DR: if we want to leverage autobuilds on dockerhub, perhaps we need a different strategy.

bam241 committed:
fixing Dockerfile with the proper path to the script

bam241 commented:
Few though there and comments:
- the `hooks/build` strategy seems to have worked to allow multiple build using a single `Dockerfile` with different `ENV_VARIABLE`
- I am not sure how the 2h limits is used, is it 2h limits per job (i.e. `hooks/build` script), per image build or per RUN command....
- it seems that the `ci/dockercloud` hook replaced our `CI/CircleCI` hook in the test check :(, not sure if we can have both into the same repo... will look into it...

bam241 committed:
add new location to be added to the dockerfile

bam241 committed:
change relative path ?

bam241 committed:
add scripts symb link

bam241 committed:
reflecting the new symb link change

bam241 committed:
new path ?

bam241 committed:
this shall be it

bam241 committed:
ls+pwd, to know where what and how

bam241 committed:
fix spelling

bam241 commented:
I am not sure what to do with this..., the dockeer build process is wait to long...

my suggestion would be to leave this on the side, maybe make a issue about this and merge this PR....

bam241 committed:
cleaning dockerhook

bam241 committed:
dummy modif to travis.yml to trigger travis

bam241 committed:
allow empty commit in travis

bam241 committed:
cleaning .travis.yml

bam241 committed:
revert some change

bam241 commented:
I am not sure how, but travis is back :)

bam241 commented:
discovered a bug in the existing `Dockerfile`, took me a couple of hours to fix it.
I am building and pushing fixed version of the docker container. 
I'll do my "final" push on this PR after :)

bam241 committed:
fixing dockerfile

bam241 commented:
@gonuke dockerhub auto-build restriction to 2h runs, will force us to redesign our workflow concerning docker image generation and usage in CI. 
(I tried to push the image before hand but still takes a super long time on dockerhub...)
After discussing with @pshriwise we concluded that it is probably better to keep this PR simple, make an issue about dockerhub auto-generation, and redesign docker image workflow in a different PR.


@ljacobson64 @kkiesling is there anything blocking still this PR for you ? (I'll be happy to discuss/tackle those points if any remain.)

I'd like to keep this moving... @pshriwise's PR #674 is waiting for this to be finalized and merged.

gonuke commented:
I think what I want to see is that the docker image generated by this process will result in an image in which testing is successful.  If you already pushed a new image based on this to dockerhub, maybe that already happened when the most recent commits were tested on Travis??  can you confirm?

bam241 commented:
Yes I pushed my last commit after having build and push the new docker images on `dockerhub`.

so the last set of Travis test were run on docker images generated with this docker image

bam241 commented:
todo:
- [x] move housekeeping the the beginning
- [x] stop CI if one test fail

bam241 committed:
re-ordering travis jobs

bam241 committed:
naming the first test jobs

bam241 committed:
messing up some formating to have housekeeping fail

bam241 committed:
revert change in DagMC.cpp

bam241 commented:
@pshriwise @gonuke 
- I added a fast fail, which will stop CI jobs if one fails. (I try to with a DagMC in which I messed up the formatting, it fails and stoped everything [see](https://travis-ci.org/github/svalinn/DAGMC/builds/687020418))
- I reput housekeeping as the first jobs, so it should run first.

bam241 commented:
@pshriwise @gonuke 
- I added a fast fail, which will stop CI jobs if one fails. (I try to with a DagMC in which I messed up the formatting, it fails and stoped everything [see](https://travis-ci.org/github/svalinn/DAGMC/builds/687020418))
- I reput housekeeping as the first job, so it should run first.


@ljacobson64 I think I addressed all your comments, could you confirm ??

ljacobson64 commented on scripts/build_geant4.sh:
```diff
@@ -1,17 +1,22 @@
-wget http://geant4.cern.ch/support/source/geant4.${geant4_version}.tar.gz
+wget https://geant4.cern.ch/support/source/geant4.${geant4_version}.tar.gz --no-check-certificate
+tar_chashum=$(sha256sum ${geant4_version}.tar.gz | cut -d' ' -f1)
+if [ $geant4_shasum != $tar_chashum ]; then
```
If we're doing the shasum thing for Geant4, would it make sense to do it for HDF5 too?

bam241 replied:
fair point !

ljacobson64 commented on .travis.yml:
```diff
@@ -7,44 +7,12 @@ notifications:
   - docker_image=svalinn/dagmc-ci-ubuntu-${UBUNTU_VERSION}
   - dagmc_build_dir=/root/build/${COMPILER}/DAGMC-moab-${MOAB_VERSION}-hdf5-${HDF5_VERSION}
-  - script_dir=${dagmc_build_dir}/DAGMC/docker
+  - script_dir=${dagmc_build_dir}/DAGMC/scripts
```
I don't think I understand why the scripts need to be in a separate directory from the Dockerfile. All of the scripts in this directory are CI-specific scripts that are only intended to be used in conjunction with the docker image.

bam241 replied:
I think I got confuse because some of the file where `travis.xxx` and some just `build` I though that the build one could be used elsewhere.
@ljacobson64 What about we put all of them in a CI folder as you suggested (including the `Dockerfile`) and add a `docker.` prefix in the build scripts that are intended to be used in the docker building process.

ljacobson64 replied:
I like that idea. So there'd be a folder called `CI` and it would contain these files

```
docker.build.dagmc.sh
docker.build.geant4.sh
docker.build.hdf5.sh
docker.build.moab.sh
docker.env.sh
Dockerfile
travis.housekeeping.sh
travis.install.sh
travis.tests.sh
update_docker.sh
```

bam241 replied:
build_dagmc is only used in travis, so I named it `travis.`
still working on making sure update_docker still work before pushing it

bam241 replied:
`env.sh` is used in both, so I kept it without a prefix

gonuke replied:
Sorry to be late here, but why `CI/travis.xxx` and `CI/docker.xxxx` and not `CI/travis/xxxx` and `CI/docker/xxxx`?

ljacobson64 commented:
One final comment: I think it would be a good idea to rebase this PR against `develop` before merging. There have been several PRs merged since this one was first submitted, so we want to make sure these new PRs also play nice with the new CI.

bam241 committed:
renaming and moving

bam241 committed:
readding push form docker_update

bam241 commented:
@ljacobson64 I added the suggestion about check_sum for hdf5 (and the moving/renaming of the scripts and Dockerfile)

the hdf5 shasum made me think about the version we are using, as when we use an other version than the system one, we only use the `1.10.4` shall we note specify it in the `env.sh` instead of having a environment variable that we set "manually" ?

bam241 committed:
changing script_dir value to CI

bam241 committed:
fix type on build_dagmc script name

ljacobson64 commented on CI/docker.build_hdf5.sh:
```diff
@@ -13,11 +13,17 @@ fi
+
 wget https://support.hdfgroup.org/ftp/HDF5/releases/hdf5-${HDF5_VERSION_major}/hdf5-${HDF5_VERSION}/src/hdf5-${HDF5_VERSION}.tar.gz
+tar_chashum=$(sha256sum hdf5-${HDF5_VERSION}.tar.gz | cut -d' ' -f1)
+if [ $hdf5_shasum != $tar_chashum ]; then
```
`$hdf5_shasum` should be `$hdf5_1_10_4_shasum`

bam241 committed:
shuffling stuff around

bam241 commented:
@ljacobson64 @gonuke I this I addressed your last comments :)

bam241 committed:
updating install script to new path

gonuke reviewed (approved):
Thanks @bam241

bam241 commented:
@ljacobson64 do you have any other comments ?

bam241 committed:
add and .vscode in gitignore

ljacobson64 commented on CI/update_docker.sh:
```diff
@@ -1,10 +1,10 @@
 for ubuntu_version in ${ubuntu_versions}; do
   image_name="svalinn/dagmc-ci-ubuntu-${ubuntu_version}:latest"
-  docker build -t ${image_name} --build-arg UBUNTU_VERSION=${ubuntu_version} .
+  docker build -t ${image_name} --build-arg UBUNTU_VERSION=${ubuntu_version} -f CI/Dockerfile .
```
Is this script now intended to be run from outside the `CI` directory?

bam241 replied:
I though it would make more sense to run from the top directory

ljacobson64 commented:
Has the docker image on dockerhub already been updated? I'm curious about how the tests are passing if it hasn't. Unless the new docker image is the same as the old one despite having a new Dockerfile.

bam241 commented:
Yes the docker container have been updated/pushed after last dockerfile change.

gonuke reviewed (approved):
Thank @bam241 - LGTM
